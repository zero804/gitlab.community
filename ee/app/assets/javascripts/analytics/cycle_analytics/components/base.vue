<script>
import { GlEmptyState } from '@gitlab/ui';
import { mapActions, mapState, mapGetters } from 'vuex';
import { PROJECTS_PER_PAGE } from '../constants';
import ProjectsDropdownFilter from '../../shared/components/projects_dropdown_filter.vue';
import { DATE_RANGE_LIMIT } from '../../shared/constants';
import DateRange from '../../shared/components/daterange.vue';
import StageTable from './stage_table.vue';
import DurationChart from './duration_chart.vue';
import TypeOfWorkCharts from './type_of_work_charts.vue';
import UrlSync from '~/vue_shared/components/url_sync.vue';
import { toYmd } from '../../shared/utils';
import StageTableNav from './stage_table_nav.vue';
import CustomStageForm from './custom_stage_form.vue';
import PathNavigation from './path_navigation.vue';
import FilterBar from './filter_bar.vue';
import ValueStreamSelect from './value_stream_select.vue';
import Metrics from './metrics.vue';

export default {
  name: 'CycleAnalytics',
  components: {
    DateRange,
    DurationChart,
    GlEmptyState,
    ProjectsDropdownFilter,
    StageTable,
    TypeOfWorkCharts,
    CustomStageForm,
    StageTableNav,
    PathNavigation,
    FilterBar,
    ValueStreamSelect,
    UrlSync,
    Metrics,
  },
  props: {
    emptyStateSvgPath: {
      type: String,
      required: true,
    },
    noDataSvgPath: {
      type: String,
      required: true,
    },
    noAccessSvgPath: {
      type: String,
      required: true,
    },
  },
  computed: {
    ...mapState([
      'featureFlags',
      'isLoading',
      'isLoadingStage',
      'isEmptyStage',
      'currentGroup',
      'selectedProjects',
      'selectedStage',
      'stages',
      'currentStageEvents',
      'errorCode',
      'startDate',
      'endDate',
      'medians',
      'isLoadingValueStreams',
      'selectedStageError',
      'selectedValueStream',
    ]),
    // NOTE: formEvents are fetched in the same request as the list of stages (fetchGroupStagesAndEvents)
    // so i think its ok to bind formEvents here even though its only used as a prop to the custom-stage-form
    ...mapState('customStages', ['isCreatingCustomStage', 'formEvents']),
    ...mapGetters([
      'hasNoAccessError',
      'currentGroupPath',
      'activeStages',
      'selectedProjectIds',
      'enableCustomOrdering',
      'cycleAnalyticsRequestParams',
      'pathNavigationData',
    ]),
    ...mapGetters('customStages', ['customStageFormActive']),
    shouldRenderEmptyState() {
      return !this.currentGroup && !this.isLoading;
    },
    shouldDisplayFilters() {
      return !this.errorCode;
    },
    shouldDisplayDurationChart() {
      return this.featureFlags.hasDurationChart && !this.hasNoAccessError;
    },
    shouldDisplayTypeOfWorkCharts() {
      return !this.hasNoAccessError;
    },
    shouldDisplayPathNavigation() {
      return this.featureFlags.hasPathNavigation && !this.hasNoAccessError && this.selectedStage;
    },
    shouldDisplayCreateMultipleValueStreams() {
      return Boolean(
        this.featureFlags.hasCreateMultipleValueStreams && !this.isLoadingValueStreams,
      );
    },
    hasDateRangeSet() {
      return this.startDate && this.endDate;
    },
    query() {
      const selectedProjectIds = this.selectedProjectIds?.length ? this.selectedProjectIds : null;

      return {
        value_stream_id: this.selectedValueStream?.id || null,
        project_ids: selectedProjectIds,
        created_after: toYmd(this.startDate),
        created_before: toYmd(this.endDate),
      };
    },
    stageCount() {
      return this.activeStages.length;
    },
    projectsQueryParams() {
      return {
        first: PROJECTS_PER_PAGE,
        includeSubgroups: true,
      };
    },
  },

  methods: {
    ...mapActions([
      'fetchCycleAnalyticsData',
      'fetchStageData',
      'setSelectedProjects',
      'setSelectedStage',
      'setDateRange',
      'removeStage',
      'updateStage',
      'reorderStage',
    ]),
    ...mapActions('customStages', ['hideForm', 'showCreateForm', 'showEditForm', 'createStage']),
    onProjectsSelect(projects) {
      this.setSelectedProjects(projects);
      this.fetchCycleAnalyticsData();
    },
    onStageSelect(stage) {
      this.hideForm();
      this.setSelectedStage(stage);
      this.fetchStageData(this.selectedStage.slug);
    },
    onShowAddStageForm() {
      this.showCreateForm();
    },
    onShowEditStageForm(initData = {}) {
      this.setSelectedStage(initData);
      this.showEditForm(initData);
    },
    onCreateCustomStage(data) {
      this.createStage(data);
    },
    onUpdateCustomStage(data) {
      this.updateStage(data);
    },
    onRemoveStage(id) {
      this.removeStage(id);
    },
    onStageReorder(data) {
      this.reorderStage(data);
    },
  },
  multiProjectSelect: true,
  dateOptions: [7, 30, 90],
  maxDateRange: DATE_RANGE_LIMIT,
};
</script>
<template>
  <div>
    <div
      class="gl-mb-3 gl-display-flex gl-flex-direction-column gl-sm-flex-direction-row gl-justify-content-space-between"
    >
      <h3>{{ __('Value Stream Analytics') }}</h3>
      <value-stream-select
        v-if="shouldDisplayCreateMultipleValueStreams"
        class="gl-align-self-start gl-sm-align-self-start gl-mt-0 gl-sm-mt-5"
        :has-path-navigation="featureFlags.hasPathNavigation"
      />
    </div>
    <gl-empty-state
      v-if="shouldRenderEmptyState"
      :title="__('Value Stream Analytics can help you determine your team’s velocity')"
      :description="
        __('Filter parameters are not valid. Make sure that the end date is after the start date.')
      "
      :svg-path="emptyStateSvgPath"
    />
    <div v-if="!shouldRenderEmptyState" class="gl-max-w-full">
      <div class="gl-mt-3 gl-py-2 gl-px-3 bg-gray-light border-top border-bottom">
        <div v-if="shouldDisplayPathNavigation" class="gl-w-full gl-pb-2">
          <path-navigation
            class="js-path-navigation"
            :loading="isLoading"
            :stages="pathNavigationData"
            :selected-stage="selectedStage"
            @selected="onStageSelect"
          />
        </div>
        <div
          v-if="shouldDisplayFilters"
          class="gl-display-flex gl-flex-direction-column gl-lg-flex-direction-row gl-justify-content-space-between"
        >
          <projects-dropdown-filter
            :key="currentGroup.id"
            class="js-projects-dropdown-filter project-select gl-mb-2 gl-lg-mb-0"
            :group-id="currentGroup.id"
            :group-namespace="currentGroupPath"
            :query-params="projectsQueryParams"
            :multi-select="$options.multiProjectSelect"
            :default-projects="selectedProjects"
            @selected="onProjectsSelect"
          />
          <date-range
            :start-date="startDate"
            :end-date="endDate"
            :max-date-range="$options.maxDateRange"
            :include-selected-date="true"
            class="js-daterange-picker"
            @change="setDateRange"
          />
        </div>
        <filter-bar
          v-if="shouldDisplayFilters"
          class="js-filter-bar filtered-search-box gl-display-flex gl-mt-3 gl-mr-3 gl-border-none"
          :group-path="currentGroupPath"
        />
      </div>
    </div>
    <div v-if="!shouldRenderEmptyState" class="cycle-analytics gl-mt-0">
      <gl-empty-state
        v-if="hasNoAccessError"
        class="js-empty-state"
        :title="__('You don’t have access to Value Stream Analytics for this group')"
        :svg-path="noAccessSvgPath"
        :description="
          __(
            'Only \'Reporter\' roles and above on tiers Premium / Silver and above can see Value Stream Analytics.',
          )
        "
      />
      <div v-else>
        <metrics :group-path="currentGroupPath" :request-params="cycleAnalyticsRequestParams" />
        <stage-table
          :key="stageCount"
          class="js-stage-table"
          :current-stage="selectedStage"
          :is-loading="isLoading"
          :is-loading-stage="isLoadingStage"
          :is-empty-stage="isEmptyStage"
          :custom-stage-form-active="customStageFormActive"
          :current-stage-events="currentStageEvents"
          :no-data-svg-path="noDataSvgPath"
          :empty-state-message="selectedStageError"
        >
          <template #nav>
            <stage-table-nav
              :current-stage="selectedStage"
              :stages="activeStages"
              :medians="medians"
              :is-creating-custom-stage="isCreatingCustomStage"
              :custom-ordering="enableCustomOrdering"
              @reorderStage="onStageReorder"
              @selectStage="onStageSelect"
              @editStage="onShowEditStageForm"
              @showAddStageForm="onShowAddStageForm"
              @hideStage="onUpdateCustomStage"
              @removeStage="onRemoveStage"
            />
          </template>
          <template v-if="customStageFormActive" #content>
            <custom-stage-form
              :events="formEvents"
              @createStage="onCreateCustomStage"
              @updateStage="onUpdateCustomStage"
              @clearErrors="$emit('clear-form-errors')"
            />
          </template>
        </stage-table>
        <url-sync :query="query" />
      </div>
      <duration-chart v-if="shouldDisplayDurationChart" class="gl-mt-3" :stages="activeStages" />
      <type-of-work-charts v-if="shouldDisplayTypeOfWorkCharts" />
    </div>
  </div>
</template>
