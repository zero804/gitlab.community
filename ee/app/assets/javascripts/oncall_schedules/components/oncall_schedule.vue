<script>
import {
  GlSprintf,
  GlCard,
  GlButtonGroup,
  GlButton,
  GlModalDirective,
  GlTooltipDirective,
} from '@gitlab/ui';
import { s__ } from '~/locale';
import ScheduleTimelineSection from './schedule/components/schedule_timeline_section.vue';
import DeleteScheduleModal from './delete_schedule_modal.vue';
import EditScheduleModal from './edit_schedule_modal.vue';
import AddEditRotationModal from './rotations/add_edit_rotation_modal.vue';

import { getTimeframeForWeeksView } from './schedule/utils';
import { PRESET_TYPES } from './schedule/constants';
import { getFormattedTimezone } from '../utils/common_utils';
import RotationsListSection from './schedule/components/rotations_list_section.vue';

import { addRotationModalId, editRotationModalId } from '../constants';

export const i18n = {
  scheduleForTz: s__('OnCallSchedules|On-call schedule for the %{tzShort}'),
  editScheduleLabel: s__('OnCallSchedules|Edit schedule'),
  deleteScheduleLabel: s__('OnCallSchedules|Delete schedule'),
  rotationTitle: s__('OnCallSchedules|Rotations'),
  addARotation: s__('OnCallSchedules|Add a rotation'),
};
export const editScheduleModalId = 'editScheduleModal';
export const deleteScheduleModalId = 'deleteScheduleModal';

export default {
  i18n,
  addRotationModalId,
  editRotationModalId,
  editScheduleModalId,
  deleteScheduleModalId,
  presetType: PRESET_TYPES.WEEKS,
  inject: ['timezones'],
  components: {
    GlSprintf,
    GlCard,
    ScheduleTimelineSection,
    GlButtonGroup,
    GlButton,
    DeleteScheduleModal,
    EditScheduleModal,
    AddEditRotationModal,
    RotationsListSection,
  },
  directives: {
    GlModal: GlModalDirective,
    GlTooltip: GlTooltipDirective,
  },
  props: {
    schedule: {
      type: Object,
      required: true,
    },
    rotations: {
      type: Array,
      required: false,
      default: () => [],
    },
  },
  computed: {
    tzLong() {
      const selectedTz = this.timezones.find(tz => tz.identifier === this.schedule.timezone);
      return getFormattedTimezone(selectedTz);
    },
    timeframe() {
      return getTimeframeForWeeksView();
    },
  },
};
</script>

<template>
  <div>
    <gl-card>
      <template #header>
        <div
          class="gl-display-flex gl-justify-content-space-between gl-m-0"
          data-testid="scheduleHeader"
        >
          <span class="gl-font-weight-bold gl-font-lg">{{ schedule.name }}</span>
          <gl-button-group>
            <gl-button
              v-gl-modal="$options.editScheduleModalId"
              v-gl-tooltip
              :title="$options.i18n.editScheduleLabel"
              icon="pencil"
              :aria-label="$options.i18n.editScheduleLabel"
            />
            <gl-button
              v-gl-modal="$options.deleteScheduleModalId"
              v-gl-tooltip
              :title="$options.i18n.deleteScheduleLabel"
              icon="remove"
              :aria-label="$options.i18n.deleteScheduleLabel"
            />
          </gl-button-group>
        </div>
      </template>
      <p class="gl-text-gray-500 gl-mb-5" data-testid="scheduleBody">
        <gl-sprintf :message="$options.i18n.scheduleForTz">
          <template #tzShort>{{ schedule.timezone }}</template>
        </gl-sprintf>
        | {{ tzLong }}
      </p>

      <gl-card header-class="gl-bg-transparent">
        <template #header>
          <div
            class="gl-display-flex gl-justify-content-space-between"
            data-testid="rotationsHeader"
          >
            <h6 class="gl-m-0">{{ $options.i18n.rotationTitle }}</h6>
            <gl-button v-gl-modal="$options.addRotationModalId" variant="link"
              >{{ $options.i18n.addARotation }}
            </gl-button>
          </div>
        </template>

        <div class="schedule-shell" data-testid="rotationsBody">
          <schedule-timeline-section :preset-type="$options.presetType" :timeframe="timeframe" />
          <rotations-list-section
            :preset-type="$options.presetType"
            :rotations="rotations"
            :timeframe="timeframe"
          />
        </div>
      </gl-card>
    </gl-card>
    <delete-schedule-modal :schedule="schedule" :modal-id="$options.deleteScheduleModalId" />
    <edit-schedule-modal :schedule="schedule" :modal-id="$options.editScheduleModalId" />
    <add-edit-rotation-modal :schedule="schedule" :modal-id="$options.addRotationModalId" />
    <add-edit-rotation-modal
      :schedule="schedule"
      :modal-id="$options.editRotationModalId"
      is-edit-mode
    />
  </div>
</template>
