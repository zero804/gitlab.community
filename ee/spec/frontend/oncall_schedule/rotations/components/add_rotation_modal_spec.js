import { shallowMount, createLocalVue } from '@vue/test-utils';
import createMockApollo from 'jest/helpers/mock_apollo_helper';
import VueApollo from 'vue-apollo';
import waitForPromises from 'helpers/wait_for_promises';
import { GlDropdownItem, GlModal, GlAlert, GlTokenSelector, GlToggle } from '@gitlab/ui';
import { addRotationModalId } from 'ee/oncall_schedules/components/oncall_schedule';
import AddRotationModal from 'ee/oncall_schedules/components/rotations/components/add_rotation_modal.vue';
// import createOncallScheduleRotationMutation from 'ee/oncall_schedules/graphql/create_oncall_schedule_rotation.mutation.graphql';
import usersSearchQuery from '~/graphql_shared/queries/users_search.query.graphql';
import { getOncallSchedulesQueryResponse, participants } from '../../mocks/apollo_mock';

const schedule =
  getOncallSchedulesQueryResponse.data.project.incidentManagementOncallSchedules.nodes[0];
const localVue = createLocalVue();
const projectPath = 'group/project';
const mutate = jest.fn();
const mockHideModal = jest.fn();

localVue.use(VueApollo);

describe('AddRotationModal', () => {
  let wrapper;
  let fakeApollo;
  let userSearchQueryHandler;

  async function awaitApolloDomMock() {
    await wrapper.vm.$nextTick(); // kick off the DOM update
    await jest.runOnlyPendingTimers(); // kick off the mocked GQL stuff (promises)
    await wrapper.vm.$nextTick(); // kick off the DOM update for flash
  }

  const createComponent = ({ data = {}, props = {}, loading = false } = {}) => {
    wrapper = shallowMount(AddRotationModal, {
      data() {
        return {
          ...data,
        };
      },
      propsData: {
        modalId: addRotationModalId,
        schedule,
        ...props,
      },
      provide: {
        projectPath,
      },
      mocks: {
        $apollo: {
          queries: {
            participants: {
              loading,
            },
          },
          mutate,
        },
      },
    });
    wrapper.vm.$refs.createScheduleRotationModal.hide = mockHideModal;
  };

  const createComponentWithApollo = ({ search = '' } = {}) => {
    fakeApollo = createMockApollo([[usersSearchQuery, userSearchQueryHandler]]);

    wrapper = shallowMount(AddRotationModal, {
      localVue,
      propsData: {
        modalId: addRotationModalId,
        schedule,
      },
      apolloProvider: fakeApollo,
      data() {
        return {
          ptSearchTerm: search,
          form: {
            participants,
          },
          participants,
        };
      },
      provide: {
        projectPath,
      },
    });
  };

  beforeEach(() => {
    createComponent();
  });

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  const findModal = () => wrapper.find(GlModal);
  const findRotationLength = () => wrapper.find('[id = "rotation-length"]');
  const findRotationStartTime = () => wrapper.find('[id = "rotation-start-time"]');
  const findRotationEndsContainer = () => wrapper.find('[data-testid = "ends-on"]');
  const finEndDateToggle = () => wrapper.find(GlToggle);
  const findRotationEndTime = () => wrapper.find('[id = "rotation-end-time"]');
  const findUserSelector = () => wrapper.find(GlTokenSelector);
  const findStartsOnTimeOptions = () => findRotationStartTime().findAll(GlDropdownItem);
  const findEndsOnTimeOptions = () => findRotationEndTime().findAll(GlDropdownItem);
  const findAlert = () => wrapper.find(GlAlert);

  it('renders rotation modal layout', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('Rotation length and start time', () => {
    it('renders the rotation length value', async () => {
      const rotationLength = findRotationLength();
      expect(rotationLength.exists()).toBe(true);
      expect(rotationLength.attributes('value')).toBe('1');
    });

    it('renders the rotation starts on datepicker', async () => {
      const startsOn = findRotationStartTime();
      expect(startsOn.exists()).toBe(true);
      expect(startsOn.attributes('text')).toBe('00:00');
      expect(startsOn.attributes('headertext')).toBe('');
    });

    it('should add a checkmark to a selected start time', async () => {
      const options = findStartsOnTimeOptions();
      const time1 = options.at(0);
      const time2 = options.at(1);
      time1.vm.$emit('click');
      await wrapper.vm.$nextTick();
      expect(time1.props('isChecked')).toBe(true);
      expect(time2.props('isChecked')).toBe(false);
    });
  });

  describe('Rotation end time', () => {
    it('toggles end time visibility', async () => {
      const toggle = finEndDateToggle().vm;
      toggle.$emit('change', false);
      await wrapper.vm.$nextTick();
      expect(findRotationEndsContainer().exists()).toBe(false);
      toggle.$emit('change', true);
      await wrapper.vm.$nextTick();
      expect(findRotationEndsContainer().exists()).toBe(true);
    });

    it('should add a checkmark to a selected end time', async () => {
      finEndDateToggle().vm.$emit('change', true);
      await wrapper.vm.$nextTick();
      const options = findEndsOnTimeOptions();
      const time1 = options.at(0);
      const time2 = options.at(1);
      time2.vm.$emit('click');
      await wrapper.vm.$nextTick();
      expect(time1.props('isChecked')).toBe(false);
      expect(time2.props('isChecked')).toBe(true);
    });
  });

  describe('filter participants', () => {
    beforeEach(() => {
      createComponent({ data: { participants } });
    });

    it('has user options that are populated via apollo', () => {
      expect(findUserSelector().props('dropdownItems').length).toBe(participants.length);
    });

    it('calls the API and sets dropdown items as request result', async () => {
      const tokenSelector = findUserSelector();

      tokenSelector.vm.$emit('focus');
      tokenSelector.vm.$emit('blur');
      tokenSelector.vm.$emit('focus');

      await waitForPromises();

      expect(tokenSelector.props('dropdownItems')).toMatchObject(participants);
      expect(tokenSelector.props('hideDropdownWithNoItems')).toBe(false);
    });

    it('emits `input` event with selected users', () => {
      findUserSelector().vm.$emit('input', participants);

      expect(findUserSelector().emitted().input[0][0]).toEqual(participants);
    });

    it('when text input is blurred the text input clears', async () => {
      const tokenSelector = findUserSelector();
      tokenSelector.vm.$emit('blur');

      await wrapper.vm.$nextTick();

      expect(tokenSelector.props('hideDropdownWithNoItems')).toBe(false);
    });
  });

  describe('Rotation create', () => {
    it('makes a request with `oncallScheduleRotationCreate` to create a schedule rotation', () => {
      mutate.mockResolvedValueOnce({});
      findModal().vm.$emit('primary', { preventDefault: jest.fn() });
      expect(mutate).toHaveBeenCalledWith({
        mutation: expect.any(Object),
        variables: { oncallScheduleRotationCreate: expect.objectContaining({ projectPath }) },
      });
    });

    it('does not hide the rotation modal and shows error alert on fail', async () => {
      const error = 'some error';
      mutate.mockResolvedValueOnce({ data: { oncallScheduleRotationCreate: { errors: [error] } } });
      findModal().vm.$emit('primary', { preventDefault: jest.fn() });
      await waitForPromises();
      expect(mockHideModal).not.toHaveBeenCalled();
      expect(findAlert().exists()).toBe(true);
      expect(findAlert().text()).toContain(error);
    });
  });

  describe('with mocked Apollo client', () => {
    it('it calls searchUsers query with the search paramter', async () => {
      userSearchQueryHandler = jest.fn().mockResolvedValue({
        data: {
          users: {
            nodes: participants,
          },
        },
      });
      createComponentWithApollo({ search: 'root' });
      await awaitApolloDomMock();
      expect(userSearchQueryHandler).toHaveBeenCalledWith({ search: 'root' });
    });

    // TODO: Once the BE is complete for the mutation add specs here for that via a creationHandler
  });
});
