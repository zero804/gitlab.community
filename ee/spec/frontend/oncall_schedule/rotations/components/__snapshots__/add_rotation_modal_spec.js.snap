// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRotationModal renders rotation modal layout 1`] = `
<gl-modal-stub
  actioncancel="[object Object]"
  actionprimary="[object Object]"
  modalclass="rotations-modal"
  modalid="addRotationModal"
  size="md"
  title="Add rotation"
  titletag="h4"
>
  <!---->
   
  <gl-form-stub>
    <div
      class="w-75 gl-xs-w-full!"
    >
      <gl-form-group-stub
        invalid-feedback="Rotation name cannot be empty"
        label="Name"
        label-for="rotation-name"
        label-size="sm"
        state="true"
      >
        <gl-form-input-stub
          id="rotation-name"
          value=""
        />
      </gl-form-group-stub>
       
      <gl-form-group-stub
        invalid-feedback="Rotation participants cannot be empty"
        label="Participants"
        label-for="rotation-participants"
        label-size="sm"
        state="true"
      >
        <gl-token-selector-stub
          autocomplete="off"
          containerclass="gl-h-13! gl-overflow-y-auto"
          dropdownitems=""
          selectedtokens=""
        />
      </gl-form-group-stub>
       
      <gl-form-group-stub
        label="Rotation length"
        label-for="rotation-length"
        label-size="sm"
      >
        <div
          class="gl-display-flex"
        >
          <gl-form-input-stub
            class="gl-w-12 gl-mr-3"
            id="rotation-length"
            min="1"
            type="number"
            value="1"
          />
           
          <gl-dropdown-stub
            category="primary"
            headertext=""
            size="medium"
            text="hours"
            variant="default"
          >
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischecked="true"
              ischeckitem="true"
              secondarytext=""
            >
              
              hours
            
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              
              days
            
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              
              weeks
            
            </gl-dropdown-item-stub>
          </gl-dropdown-stub>
        </div>
      </gl-form-group-stub>
       
      <gl-form-group-stub
        invalid-feedback="Rotation start date cannot be empty"
        label="Starts on"
        label-for="rotation-start-time"
        label-size="sm"
        state="true"
      >
        <div
          class="gl-display-flex gl-align-items-center"
        >
          <gl-datepicker-stub
            ariallabel=""
            autocomplete=""
            class="gl-mr-3"
            container=""
            displayfield="true"
            firstday="0"
            placeholder="YYYY-MM-DD"
            target=""
            theme=""
          />
           
          <span>
             at 
          </span>
           
          <gl-dropdown-stub
            category="primary"
            class="gl-w-12 gl-pl-3"
            headertext=""
            id="rotation-start-time"
            size="medium"
            text="00:00"
            variant="default"
          >
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 01:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 02:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 03:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 04:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 05:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 06:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 07:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 08:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 09:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 10:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 11:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 12:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 13:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 14:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 15:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 16:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 17:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 18:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 19:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 20:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 21:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 22:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 23:00
              </span>
            </gl-dropdown-item-stub>
            <gl-dropdown-item-stub
              avatarurl=""
              iconcolor=""
              iconname=""
              iconrightarialabel=""
              iconrightname=""
              ischeckitem="true"
              secondarytext=""
            >
              <span
                class="gl-white-space-nowrap"
              >
                 24:00
              </span>
            </gl-dropdown-item-stub>
          </gl-dropdown-stub>
           
          <span
            class="gl-pl-5"
          >
             {
  "identifier": "Pacific/Honolulu"
} 
          </span>
        </div>
      </gl-form-group-stub>
    </div>
     
    <gl-toggle-stub
      class="gl-mb-5"
      label="Enable end date"
      labelposition="left"
    />
     
    <!---->
  </gl-form-stub>
</gl-modal-stub>
`;
